### DISCUSSION TITLE
Actor syntax

### DISCUSSION THREAD

**Discussion by oleganza** [tact-lang/spec#10] [Ideas]
Cxmog suggested using comptime functions to convert ordinary interface definitions into Actor sender/recipient wrappers. Let's discuss this idea in detail in here. 

Our requirements are:

* Ability for precise message encoding definition
* Nice per-message handling syntax and understandable message-sending syntax
* First-class support for Tact data types to avoid boilerplate
* Ability to send msg with flags and values, alongside data types
* Clear distinction between external/internal messages (whatever shape that takes)
* Ability to strongly type `msg sender` to ensure cross-contract interface conformance.
* (In the future) Gas and value bounds via some form of const-generics / dependent types.

Some ideas in this regard:

**Do we allow multi-argument / single-argument payload?**

MA: more natural to express, but can't be mixed with system args (value, sender, etc.), less obvious how to customize serialization

SA: allows placing all serialization logic on one type, allows mixing with system args, but extra boilerplate - new wrapper type for each msg.

**Distinguishing external/Internal messages**

Options:

* Entirely separate syntax for message handlers: `internal foo(...) { ... }`
* Attributes: `@internal fn foo(...) {  }`
* Lack/presence of certain arguments: value/sender
* By type of Sender: `fn handler(sender: External)`

**Comment by cmxog** [comment]
To make it easier to continue down the road of figuring out exactly how actors will *look like* in code, both on sending and receiving sides, we have provisionally agreed on the using assumption that this will be based on compile-time construction as opposed to special syntax for this purpose. This way we can explore the space and figure out if there's anything wrong with it.

Provided is the early draft for the approach (with some handwaving) that I wrote a few days ago. Some further discussions already happened around it and there's going to be more effort to define this surface with better fidelity.

---

This actor proposal tries to make the use of Tact's meta-programming to avoid
extending the core language.

The core idea lies in defining an arbitrary interface for an actor. This is what
is used by both ends to agree on what is being communicated.

```
interface MyContract {

   fn approve(approval: Approval) -> Void
   fn reject(rejection: Rejection) -> Void

} 
```

(Digression: if the parser will permit, we can look into shorthanding `-> Void` into nothing)

While we *do* define the interface as function calls, they must be serializable to
be sent.

Therefore, we need to have a function that takes an interface kind (an example above is an instance of
such a kind) and, if it is valid (say, all functions return `Void`), turns it into a union of messages:

```
let MyContractMsg = contractMessage(MyContract);
```

which will essentially make `MyContractMsg` equal to this union:

```
type {approval: Approval} | type{rejection: Rejection}
```

However, most of the time, this is not to be used directly. In most cases, we can take it further and
introduce two more compile-time functions.

One would take `MyContract` and transform it into an interface that handles messages:

```
let MyContractImplementor = contractImplementationInterface(MyContract);
```

which will essentially make `MyContractImplementor` equal to this interface:

```
interface {

   fn handleMessage(self: Self, message: Message) -> Void {
       ...
       self.setSender(message.sender);
       ...
       // figure out based on serialization info of the union,
       // which function (`approve` or `reject`) to dispatch to
   }

   fn setSender(self: Self, sender: Sender) -> Void

   fn approve(self: Self, approval: Approval) -> Void
   fn reject(self: Self, rejection: Rejection) -> Void
}
```

(Please note that the details of messages are glossed over for illustrative
purposes at the moment. We can improve this proposal with real structures to
make it more tangible if the direction of thought is deemed interesting enough.)

This interface is to be implemented by a type and called from an entry point.

Now, the second function would do the caller-side generation from the same original interface:

```
let MyContractCaller = contractCaller(MyContract);
```

where `MyContract` will be equal to a type:

```
type {
   recipient: Recipient
  
   fn approve(approval: Approval, gas: Option(Gas) = None) -> Void {
        send(recipient, (unionVariant(contactMessage(MyContract), 0) {approval: Approval}).serialize()));
   }
   fn reject(rejection: Rejection) -> Void {
        send(recipient, (unionVariant(contactMessage(MyContract), 1) {rejection: Rejection}).serialize()));
   }

}
```

Where `unionVariant` is a fictional function (for now, at least -- and it may take a radically different form) to extract
the type variant and `send` is a less fictional function that roughly maps to `SENDRAWMSG`. `serialize()` is an approximation
to whatever the serialization framework will look like.


```
// ...
let caller = MyContractCaller{recipient: somebody, gas: Gas(10)};
caller.reject(rejection, gas: );
caller.approve(approval);
```
